lib/libc/minimal/include/time.h

```diff
@@ -57,11 +57,11 @@ struct tm *gmtime_r(const time_t *_MLIBC_RESTRICT timep,
 /*
  * The asctime/_r functions shall convert the broken-down time in the
  * structure pointed to by timeptr into a string in the form:
- * 
+ *
  *            Day Mon dd hh:mm:ss yyyy\n\0
  */
 char *asctime(const struct tm *timeptr);
-char *__asctime_r (const struct tm *timeptr, char *buf);
+char *asctime_r(const struct tm *timeptr, char *buf);
```

Comment:
Squash this into the first commit.

Do **not** open a new PR for this.  If you need help with using git, ask.

Reply:
Can you please guide me here? I finished the changes for now and want to make sure I will do this part good

Reply:
OK, here are some instructions based on what's currently pushed, and assuming you're using a commandline git interface (if you're using something else, I can't help you).

<details><summary>Expand for some notes</summary>

Here are your commits.
```
* 0655567cd9 (HEAD -> pr/31384) [Mon Jan 18 01:03:55 2021 +0200] lib/libc/minimal: add ignore flag for new typedefs in code
* 920f28f0b6 [Mon Jan 18 00:14:25 2021 +0200] lib/libc/minimal: coding-style errors and fix asctime_r function name
* c70feceb1e [Sun Jan 17 23:33:42 2021 +0200] first commit: add div and asctime impl
```

I want the following:

First, remove the top commit.  First, the description is wrong, as it affects commits throughout the tree, not just ones in lib/libc/minimal.  Second, it overrides a policy decision, so has to be handled separately.  It can be removed by discarding the commit entirely by resetting the head and the workspace to the previous commit.

    git reset --hard HEAD^

Now you have these commits:

```
* 920f28f0b6 (HEAD -> pr/31384) [Mon Jan 18 00:14:25 2021 +0200] lib/libc/minimal: coding-style errors and fix asctime_r function name
* c70feceb1e [Sun Jan 17 23:33:42 2021 +0200] first commit: add div and asctime impl
```

The top commit here has two changes, the removal of the include which I don't want done, and two changes to the `time.h` header that need to be done in the first commit.

First, remove the commit but leave the changes in your workspace:

    git reset --soft HEAD^
    
```
tirzah[1589]$ git status -uno
On branch pr/31384
Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
	modified:   lib/libc/minimal/include/time.h
	modified:   lib/libc/minimal/source/time/asctime.c

Untracked files not listed (use -u option to show untracked files)
```

Then take those out of the staging area:

    git reset

```
tirzah[1592]$ git status -uno
On branch pr/31384
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   lib/libc/minimal/include/time.h
	modified:   lib/libc/minimal/source/time/asctime.c
```

Verify that the asctime.c changes are just the include file:

    git diff lib/libc/minimal/source/time/asctime.c

then discard them:

    git checkout lib/libc/minimal/source/time/asctime.c

Now the only changes are to time.h.  Use diff on the workspace to see the change to the whitespace in the commits, and the renamed function declaration:

    git diff

Now amend the first commit so all the changes are in it:

    git commit --amend -C HEAD -a

Now you have one commit with all your changes.

Unfortunately, having done this, I can tell there's more to do.  The addition of the div support and its test, and the addition of the asctime implementation and its test, also need to be in separate commits.  Also the define for the length of the asctime buffer does not belong in the `time.h` header: it's not part of the C standard library.  So you need to pick up the pieces and commit them separately.

For that you'd want to run:

    git reset --soft HEAD^
    git reset
    git add -i

to get the changes into your workspace without staging them, and then incrementally add pieces.

Then the commit messages need to be cleaned up a lot.
</details>

I'm sorry, but there's a lot to fix here, and describing all the steps is difficult.  I went through the steps above, then a lot of other steps, and ended up with https://github.com/pabigot/zephyr/commits/pr/31384 which is close to what I would like this to look like: two commits, one for each feature with test, fixes incorporated.

I'd suggest experimenting in your workspace to see if you can make what you have look like that.  Look at the documentation for git reset and add and other subcommands for more help.

Reply:
Hi, im about to finish everything you guided me to do.
but I have 2 questions, first, why you don't want me to remove the include to restrict.h file from the asctime.c? I added this by mistake anyway..
The second questions is what is the right way to contribute to this project without squash my commit all the time, I mean, if im going to do it in a big contribute im about to lose the option to go back in history because all the commits just squash together, no?

Reply:
> but I have 2 questions, first, why you don't want me to remove the include to restrict.h file from the asctime.c? I added this by mistake anyway..

I mis-read something, and thought this was something that was present before your patch. If you added it by mistake, then yes remove it.

> The second questions is what is the right way to contribute to this project without squash my commit all the time, I mean, if im going to do it in a big contribute im about to lose the option to go back in history because all the commits just squash together, no?

There's some explanation in the [contribution documentation](https://docs.zephyrproject.org/latest/contribute/index.html#contribution-workflow), but basically we want each commit to be close to self-contained as possible while only addressing one item at a time.

The version you have now:
* adds both asctime and div in a single commit.  They're different  features, so they should be separate.  Some reviewers want the tests in a separate commit; unless they're large and new I personally prefer them in the commit that provides/changes whatever the test uses.  I don't think we have a clear policy on this, but I maybe just can't find it.
* has a second commit that fixes problems in the first one.  This confuses and annoys reviewers who look at the changes one commit at a time, notice the problem and write a comment about it, then go on to the next commit where it's fixed making the comment irrelevant or incorrect.
* has a third commit that is a policy change that doesn't belong in a PR that is adding functionality to minimal libc.

In short we want to be able to look at each commit and verify that it is complete and does what it's supposed to.  So that's the sort of PR you should supply.    When the code is merged we don't need to see your development history, just the final accepted version.  If you want to track history you can keep branches or tags for earlier revisions in your local development workspace.

Reply:
Thank you for your answers,I work in a place where we always look only in the last commit of a feature branch to see if we can merge it. (of course each branch is only for one feature which is not the case in my original pull request and its my fault).

If for example i would only have added the div feature, and lets say i would do couple of commit, why would you bother yourself to look at each of the commits separately? why not just to review the last one?

Reply:
and we merge the code to the master with squah of course.

Reply:
> why not just to review the last one?

Because we *don't* squash the PR when merged.  Some PRs require as many as 40 commits to carefully build to the completion of the feature they add.  With changes of that size it's critical that each one be as self-contained and as small as possible, but no smaller.